uaclient/daemon/retry_auto_attach.py

```diff
@@ -0,0 +1,228 @@
+        return '"{}"'.format(e.msg)
+    else:
+        LOG.error("Unexpected exception: {}".format(e))
+        return str(e) or "an unknown error"
+
+
+def _try_full_auto_attach():
+    persisted_options = OPTIONS_FILE.read()
+    for i in range(2):
+        options = FullAutoAttachOptions()
+        if persisted_options is not None:
+            options.enable = persisted_options.enable
+            options.enable_beta = persisted_options.enable_beta
+        try:
+            full_auto_attach(options)
+            return
+        except (
+            api_exceptions.BetaServiceError,
+            api_exceptions.EntitlementNotFoundError,
+            api_exceptions.IncompatibleEntitlementsDetected,
+        ) as e:
+            LOG.warning(
+                "auto attach failed because of args. error message: {}".format(
+                    e.msg
+                )
+            )
+            if persisted_options is not None:
+                LOG.info("trying again with defaults")
+                persisted_options = None
+            else:
+                raise e
+
+
+def cleanup(cfg: UAConfig):
+    STATE_FILE.delete()
+    OPTIONS_FILE.delete()
+    system.remove_file(AUTO_ATTACH_STATUS_MOTD_FILE)
+    cfg.notice_file.remove("", messages.AUTO_ATTACH_RETRY_NOTICE_PREFIX)
+
+
+def retry_auto_attach(cfg: UAConfig) -> None:
+    # in case we got started while already attached somehow
+    if cfg.is_attached:
+        return
+
+    # pick up where we left off
+    persisted_state = STATE_FILE.read()
+    if persisted_state is not None:
+        # skip intervals we've already waited
+        offset = persisted_state.interval_index
+        intervals = RETRY_INTERVALS[offset:]
+        failure_reason = persisted_state.failure_reason
+    else:
+        offset = 0
+        intervals = RETRY_INTERVALS
+        failure_reason = None
+
+    print(intervals)
+    for index, interval in enumerate(intervals):
+        last_attempt = datetime.datetime.now(datetime.timezone.utc)
+        next_attempt = last_attempt + datetime.timedelta(seconds=interval)
+        next_attempt = next_attempt.replace(second=0, microsecond=0)
+        STATE_FILE.write(
+            RetryState(
+                interval_index=offset + index,
+                failure_reason=failure_reason,
+            )
+        )
+        msg_reason = failure_reason
+        if msg_reason is None:
+            msg_reason = "an unknown error"
+        try:
+            next_attempt = next_attempt.astimezone()
+        except Exception:
+            pass
+        auto_attach_status_msg = messages.AUTO_ATTACH_RETRY_NOTICE.format(
+            num_attempts=offset + index + 1,
+            reason=msg_reason,
+            next_run_datestring=next_attempt.isoformat(),
+        )
+        system.write_file(
+            AUTO_ATTACH_STATUS_MOTD_FILE, auto_attach_status_msg + "\n\n"
+        )
+        cfg.notice_file.remove("", messages.AUTO_ATTACH_RETRY_NOTICE_PREFIX)
```

Comment:
Should we consider acquiring a lock to write the notices file ? I know we are doing this on other service we have, but as we are adding more and more parallel jobs, we could start to see odd states in the future

Reply:
Yeah, doesn't hurt to grab it with a spin lock, and we can just `pass` on a LockHeldError